### ISSUE TITLE
gfx_vertex_struct! macro fails if doc comments are added

### ISSUE THREAD

**Issue by ebkalderon** [gfx-rs/gfx#1176]
Hello! I have a crate which has `#![deny(missing_docs)]` enabled, which requires that doc comments are provided for all functions, structs, fields, traits, and modules. If I attempt to build the crate, the compiler complains because I didn't comment the fields of the macro:

```
error: missing documentation for a struct field
  --> renderer/src/lib.rs:45:1
   |
45 |   gfx_vertex_struct!(VertexPosNormal {
   |  _^ starting here...
46 | |     pos: [f32; 3] = "a_Pos",
47 | |     normal: [f32; 3] = "a_Normal",
48 | |     tex_coord: [f32; 2] = "a_TexCoord",
49 | | });
   | |___^ ...ending here
   |
   = note: this error originates in a macro outside of the current crate
```

When I tried to compile the following Rust code on the latest nightly (currently 1.16.0):

```rust
gfx_vertex_struct!(VertexPosNormal {                                                
    /// Position of the vertex in three-dimensional space.                      
    pos: [f32; 3] = "a_Pos",                                                    
    /// Geometric normal vector of the vertex.                                  
    normal: [f32; 3] = "a_Normal",                                              
    /// UV texture coordinates of the vertex.                                   
    tex_coord: [f32; 2] = "a_TexCoord",                                         
});
```

I receive the following error:

```
error: expected ident, found #
  --> renderer/src/lib.rs:46:5
   |
46 |     /// Position of the vertex in three-dimensional space.
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I am using `gfx` version 0.12 straight from Crates.io. This is a massive problem as I cannot compile my crate at all as long as `gfx` is a dependency and I require documentation of all fields. Any clue what might be going wrong here?

**Comment by kvark** [comment]
Could you try doing:
```rust
#[allow(missing_docs)]
gfx_vertex_struct!(...)
```

**Comment by ebkalderon** [comment]
I tried it, and it still gives me a ton of errors because the `#[allow(missing_docs)]` only applies to that one line and not to the entire expanded code. 😕

**Comment by kvark** [comment]
Hmm, it is supposed to affect the whole item, which is the struct definition.
Similar case - https://github.com/gfx-rs/gfx/blob/master/src/render/src/pso/mod.rs#L61

**Comment by ebkalderon** [comment]
I haven't looked at the macro definition myself. Does this macro also have `impl` methods or other hidden structs/fields inside of it? If so, that might make sense. If not, this may potentially be a compiler bug.

Which gets evaluated first by the compiler, the `#` directives or `macro_rules!`?

**Comment by kvark** [comment]
This particular macro is just supposed to declare a struct and then implement a trait for it, so it should work.
However, since it doesn't work, you can do another workaround - have a specific module for all `gfx_*` macro usage, and disable the comment restriction *only* inside this whole module.

**Comment by ebkalderon** [comment]
@kvark Okay, I've checked the macro definition and think I know why it doesn't work. There are actually two traits being implemented on this struct: `Pod` and `Structure`. `Pod` has no methods, and therefore doesn't violate the `#![deny(missing_docs)]` directive. However, the `Structure` trait has a single `query()` method which does *not* have documentation, and that must be what is tripping the compiler up.

If I understand correctly, the macro below:

```rust
#![allow(missing_docs)]
gfx_vertex_struct!(VertexPosNormal {
    // Some random fields.
});
```

would essentially expand to:

```rust
#![allow(missing_docs)]
#[derive(Clone, Copy, Debug)]
#[some_other_attributes_here]
struct VertexPosNormal {
    // The fields go here. The #![allow(missing_docs)] directive is right
    // above, so the compiler moves on without complaining.
}

// This is a trait impl without methods, therefore doesn't need API docs.
// Again, the compiler moves on without complaining.
unsafe impl gfx::traits::Pod for VertexPosNormal {}

impl gfx::pso::buffer::Structure<$runtime_format> for VertexPosNormal {
    fn query(name: &str) -> Option<gfx::pso::buffer::Element<$runtime_format>> {
        // ERROR! This method is missing API documentation, and there
        // is no `#![allow(missing_docs)] directive to be found.
    }
}
```

If the macro were to enclose all that stuff into an inline `mod gfx_VertexPosNormal {  }` and exported its contents outside, then it might be fine. Or alternatively, maybe GFX should just insert a `#![allow(missing_docs)]` right above `query()` as a simple workaround. 😛

**Comment by ebkalderon** [comment]
Interestingly, the macro only fails to parse this piece with doc comments:
```rust
/// This is a vertex struct.
gfx_vertex_struct!(VertexPosNormal {                                                
    /// Position of the vertex in three-dimensional space.                      
    pos: [f32; 3] = "a_Pos",                                                    
    /// Geometric normal vector of the vertex.                                  
    normal: [f32; 3] = "a_Normal",                                              
    /// UV texture coordinates of the vertex.                                   
    tex_coord: [f32; 2] = "a_TexCoord",                                         
});
```
but not this with normal double-slash comments:
```rust
/// This is a vertex struct.
gfx_vertex_struct!(VertexPosNormal {                                                
    // Position of the vertex in three-dimensional space.                      
    pos: [f32; 3] = "a_Pos",                                                    
    // Geometric normal vector of the vertex.                                  
    normal: [f32; 3] = "a_Normal",                                              
    // UV texture coordinates of the vertex.                                   
    tex_coord: [f32; 2] = "a_TexCoord",                                         
});
```
I may be asking a silly question, but is the macro system supposed to behave this way?

**Comment by Bastacyclop** [comment]
I think that doc comments need to be handled by the macro because they have to be inserted in the macro output and not to the macro input, while simple comments can just be removed from the macro input.

**Comment by kvark** [comment]
@ebkalderon implementing a trait does not require any doc comments. I think what happens is - putting an attribute on top of the macro simply does not end up being applied to the generated struct in the output.
Hence, I'd accept a PR that either puts `#[allow(missing_docs)]` on top of all generated structs/modules, or parses the doc comments somehow.

You can still around it by having a non-documented module containing those macro invocations.

**Comment by ebkalderon** [comment]
Sounds good to me! I will make a pull request with the `#[allow(missing_docs)]` added to the macro definitions.

**Comment by ebkalderon** [comment]
Fixed by #1179!
